pkg/leaderelection/leaderelection.go

```diff
@@ -0,0 +1,203 @@
+package leaderelection
+
+import (
+	"context"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/runtime"
+
+	"github.com/razorpay/metro/pkg/logger"
+
+	"github.com/razorpay/metro/pkg/registry"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+const (
+	// JitterFactor used in wait utils
+	JitterFactor = 1.2
+)
+
+// NewLeaderElector creates a LeaderElector from a LeaderElection Config
+func NewLeaderElector(c Config, registry registry.Registry) (*LeaderElector, error) {
+	if err := c.Validate(); err != nil {
+		return nil, err
+	}
+
+	le := LeaderElector{
+		config:   c,
+		registry: registry,
+		nodeID:   "",
+	}
+
+	return &le, nil
+}
+
+// LeaderElector is a leader election client.
+type LeaderElector struct {
+	// nodeID represents the id assigned by registry
+	nodeID string
+
+	// leaderID stores id the current leader
+	leaderID string
+
+	// Leader election config
+	config Config
+
+	// registry being used for leader election
+	registry registry.Registry
+}
+
+// Run starts the leader election loop. Run will not return
+// before leader election loop is stopped by ctx or it has
+// stopped holding the leader lease
+func (le *LeaderElector) Run(ctx context.Context) {
+	defer runtime.HandleCrash()
+	defer func() {
+		le.config.Callbacks.OnStoppedLeading()
+	}()
+
+	if !le.acquire(ctx) {
+		return // ctx signalled done
+	}
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	go le.config.Callbacks.OnStartedLeading(ctx)
+	le.renew(ctx)
```

Comment:
nit: leaderelection and registry both have `renew` functions. Have different names?

Reply:
if this function returns, it would silently terminate the goroutine. The application should know that this has returned and probably shutdown the app?

Reply:
adding a TODO for this too, had thought of this, didn't spent time on solving